### ISSUE TITLE
renderUI creates multiple UI elements

### ISSUE THREAD

**Issue by fbreitwieser** [rstudio/shiny#1168]
I expected that when calling renderUI multiple times, it would overwrite the previous instances, and thus effectively update the instance. In fact it overplots it.

Is this a bug or expected behaviour? If it is expected, what is the best way to delete the UI element beforehand?

Use case: I change the height of a d3heatmap based on the number of elements there are. Thus I return a d3heatmapOutput in the renderUI.

**Comment by bborgesr** [comment]
Can you post a minimal example, so we can reproduce it? It seems you might be assigning `output$yourUI` repeatedly, which will not work very well. If, however, your `renderUI` just appears once and the height is computed there (taking some dependency on `input$nElements` or something to that effect), then it should overwrite... [To be clear, *you* should not be calling `renderUI` multiple times - you should create a single `renderUI` that "knows" when it's supposed to get re-executed because its dependencies have changed.] So, if that's the case, I'd need to look at your code more closely to understand what is going wrong.

**Comment by fbreitwieser** [comment]
Thanks for the quick reply, @bborgesr ! I extracted a minimal example, and it works just as expected.

See https://gist.github.com/fbreitwieser/c08710f1a4a13aab34d54e1609a8516f

In my application I'm using it inside a shinydashboard. I'll create a minimal example with a dashboard to see if that could be the cause.

**Comment by fbreitwieser** [comment]
Example how it screws up in my application: 
![image](https://cloud.githubusercontent.com/assets/516060/14685846/5557bc26-0704-11e6-9951-ae2ffe015401.png)

The rows of the initial heatmap were Eukaryota, Bacteria, Viruses, Archaea - they are still visible. 

As the minimal examples work, it seems to be a bug in my application. I'll investigate and post once I found a reason.

**Comment by wch** [comment]
Just to be clear: are you saying that the examples you've provided do NOT reproduce the bug?

**Comment by fbreitwieser** [comment]
Exactly, those examples work as expected and do not show the issue I am having. 

I think what I'm experiencing might be due to using data from a server-side DT that is on another tab. I'll generate one more example that reflects that - maybe that reproduces the bug.

**Comment by fbreitwieser** [comment]
Below's part of my actual code. Do you see anything I did not reflect in the example?

So it goes wron in the following way:
 - first heatmap looks just fine
 - when I then change something in the table, e.g. display more rows, and go back to the heatmap, it overplots a resized heatmap with the previous data and the current data
 - I can fix this by resizing the window - then only the heatmap with the current data remains

```
  summarized_report <- reactive({
    get_summarized_report(input$classification_level, input$contaminant_selector3, input$numeric_display)
  })

  output$dt_samples_comparison <- DT::renderDataTable({
    datatable(summarized_report())
  })


  summarized_report_matrix <- reactive({
    get_summarized_report(input$classification_level, input$contaminant_selector3, input$numeric_display, as_matrix=TRUE)
  })

  output$d3heatmap_samples_comparison <- renderUI({
    req(input$dt_samples_comparison_rows_current)
    d3heatmap::d3heatmapOutput('my_d3heatmap',width="100%",
                               height=paste0(200+length(input$dt_samples_comparison_rows_current)*15,"px"))
  })

  output$my_d3heatmap <- d3heatmap::renderD3heatmap({
    req(input$dt_samples_comparison_rows_current)
    library(d3heatmap)
    report_mat <- summarized_report_matrix()
    report_mat <- zero_if_na(report_mat[input$dt_samples_comparison_rows_current,])
    report_mat[report_mat < 0] <- 0
    d3heatmap::d3heatmap(report_mat,
                         Rowv="row" %in% input$heatmap_cluster,
                         Colv="column" %in% input$heatmap_cluster, # No Column reordering
                         scale=input$heatmap_scale,
                         colors=colorRampPalette(c("blue", "white", "red"))(100)
                         )
  })

```

**Comment by wch** [comment]
Are the table and heatmap displaying in different tab panels?

**Comment by fbreitwieser** [comment]
Yes, and this causes the `dt_rows_current` to be not necessarily up to date. I think could circumvent this by focusing the table tab when the input changes.

**Comment by wch** [comment]
I suspect it has something to do with d3heatmap not properly updating when the heatmap is not visible. It would help to have an example that exhibits the problem.

**Comment by fbreitwieser** [comment]
I found the problem! Not really the source for it, but at least a solution :). 

For the heatmap, I called the function that gathers the data again, with the `as_matrix=T` parameter. Once I changed it such that I would take the same reactive value as the datatable, the glitch did not appear anymore. It seems the wait for the recalculation caused the problem? I added a Sys.sleep to the gist to try to reproduce the problem, but still couldn't.

Also, by the way, I found out the problem appeared only with Firefox, but not with chromium. 

Thanks for your help and great responsiveness!
